draft-ietf-tls-tls13.md

```diff
@@ -1051,14 +1047,22 @@ fragment
 : The AEAD encrypted form of TLSPlaintext.fragment.
 {:br }
 
-Each AEAD cipher suite MUST specify how the nonce supplied to the AEAD
-operation is constructed, and what is the length of the
-TLSCiphertext.nonce_explicit part. In many cases, it is appropriate to use
-the partially implicit nonce technique described in Section 3.2.1 of
-{{RFC5116}}; with record_iv_length being the length of the explicit part. In
-this case, the implicit part SHOULD be derived from key_block as
-client_write_iv and server_write_iv (as described in {{key-calculation}}), and
-the explicit part is included in GenericAEAEDCipher.nonce_explicit.
+
+The nonce for the AEAD construction is formed as follows:
+
+  1. iv_length is set to max(8 bytes, N_MIN) for the AEAD algorithm
+     (see {{RFC5116}} Section 4).
+
+  2. The 64-bit sequence number is padded to the left with zeroes
+     to iv_length and XORed with the fixed iv, which is of length
+     iv_length.
+
+The resulting quantity (of length iv_length) is used as the per-record
+nonce. An AEAD algorithm where N_MAX is less than 64 bits MUST not be
+used with TLS.
```

Comment:
I think it needs to be written more clearly that this is an action done for each record. Every time, the iv needs XORing with the seq num, as that's what's incremented. It's stated earlier that the seq num may not rollover at 0xffffffffffffffff, and incrementing and rolling over the per-record nonce would be something that might get screwed up, not to mention the added check to make sure you don't go past the starting point (iv).

I'd separate out determination of iv_length from point 1 above and move the last sentence above here to the end of it. (also N_MAX is in bytes, sot the 64-bit requirement should probably be 8 bytes like N_MIN) Then:
"
The nonce for each record's AEAD construction is formed as follows:
1. The 64-bit sequence number is padded to the left with zeros up to iv_length.
2. The padded sequence number is XORed with the client or server write iv, depending on role.
"

I shall use the possibility that I might've misunderstood something here as potential evidence that it needs to be written out more explicitly so as to avoid misunderstanding. ;)